pkg/destroy/openstack/openstack.go

```diff
@@ -136,15 +136,42 @@ func (o *ClusterUninstaller) Run() error {
 }
 
 func deleteRunner(deleteFuncName string, dFunction deleteFunc, opts *clientconfig.ClientOpts, filter Filter, logger logrus.FieldLogger, channel chan string) {
-	backoffSettings := wait.Backoff{
-		Duration: time.Second * 15,
-		Factor:   1.3,
-		Steps:    25,
-	}
+	var err error
+	var gerr409 gophercloud.ErrDefault409
+	duration := 15.0
+	factor := 1.3
+
+	// This is the maximum number of attempts we can repeat in the case of regular retries.
+	maxRegularSteps := 10
+	// This is the maximum number of attempts we can repeat in the case of Conflict(409) errors retries.
+	maxExponentialBackoffSteps := 25
+
+	regularStepsCounter := 0
+	exponentialBackoffStepsCounter := 0
+
+	for {
+		finished, err := dFunction(opts, filter, logger)
+		if finished {
+			break
+		}
 
-	err := wait.ExponentialBackoff(backoffSettings, func() (bool, error) {
-		return dFunction(opts, filter, logger)
-	})
+		// If we have a Conflict, then add exponential sleeping time, otherwise retry immediately.
+		if errors.As(err, &gerr409) {
+			logger.Debugf("Retry %v with exponential backoff because of the error: %v. Waiting time before the next call: %v seconds", deleteFuncName, err, int(duration))
+			time.Sleep(time.Duration(int(duration)) * time.Second)
+			duration *= factor
+			exponentialBackoffStepsCounter++
```

Comment:
I suggest that you use `wait.ExponentialBackoff` rather than roll your own exponential backoff.

```
	backoff := wait.Backoff{
		Duration: 15 * time.Second,
		Factor:   1.3,
		Steps:    25,
	}
	wait.ExponentialBackoff(backoff, func() (bool, error) {
		for ; regularStepsCounter < maxRegularSteps; regularStepsCounter++ {
			finished, err := dFunction(opts, filter, logger)
			if finished {
				return true, err
			}

			// If we have a Conflict, then add exponential sleeping time, otherwise retry immediately.
			if errors.As(err, &gerr409) {
				logger.Debugf("Retry %v with exponential backoff because of the error: %v.", deleteFuncName, err)
				return false, nil
			}
			
			logger.Debugf("Retry %v because of the error: %v", deleteFuncName, err)
		}
		return false, wait.ErrWaitTimeout
	})
```

Reply:
looks good! it slightly modifies the logic and with this we will have 10 regular attempts for each Conflict retry. But I believe it is mostly the same

Reply:
Yes, the way that you wrote it you will have 10 regular attempts for each conflict retry. If you want to keep the old behavior, then keep the `stepsCounter` variable out of the condition function.